doc/man3/OPENSSL_version_text.pod

```diff
@@ -0,0 +1,204 @@
+C<"-dev"> for an ongoing development snapshot, C<"-alpha3"> for an
+alpha release, etc...
+The value must be a string.
+
+The macro B<OPENSSL_VERSION_BUILD_METADATA> is extra metadata, reserved
+for other parties (examples: C<"+fips">, C<"+vendor.1">).
+The OpenSSL project will not touch this macro.
+The value must be a string.
+
+=head2 Functions
+
+OPENSSL_version_major(), OPENSSL_version_minor(), OPENSSL_version_patch(),
+OPENSSL_version_pre_release(), and OPENSSL_version_build_metadata() return
+the values of the macros above for the build of the library, respectively.
+
+OPENSSL_version_text() returns the version number "MAJOR.MINOR.PATCH" where
+MINOR, MAJOR and PATCH are the values of B<OPENSSL_VERSION_MAJOR>,
+B<OPENSSL_VERSION_MINOR>, and B<OPENSSL_VERSION_PATCH>, respectively.
+
+OpenSSL_version_text_full() returns the version number text that
+OpenSSL_version_text() returns, with the strings of
+B<OPENSSL_VERSION_PRE_RELEASE> and B<OPENSSL_VERSION_BUILD_METADATA> when
+these are each defined.
+
+OPENSSL_info() returns different strings depending on B<t>:
+
+=over 4
+
+=item OPENSSL_CFLAGS
+
+The compiler flags set for the compilation process in the form
+"compiler: ..."  if available or "compiler: information not available"
+otherwise.
+
+=item OPENSSL_BUILT_ON
+
+The date of the build process in the form "built on: ..." if available
+or "built on: date not available" otherwise.
+
+=item OPENSSL_PLATFORM
+
+The "Configure" target of the library build in the form "platform: ..."
+if available or "platform: information not available" otherwise.
+
+=item OPENSSL_DIR
+
+The "OPENSSLDIR" setting of the library build in the form "OPENSSLDIR: "...""
+if available or "OPENSSLDIR: N/A" otherwise.
+
+=item OPENSSL_ENGINES_DIR
+
+The "ENGINESDIR" setting of the library build in the form "ENGINESDIR: "...""
+if available or "ENGINESDIR: N/A" otherwise.
+
+=back
+
+For an unknown B<t>, the text "not available" is returned.
+
+=head1 BACKWARD COMPATIBILITY
+
+For compatibility, some older macros and functions are retained or
+synthesised.
+They are all considered deprecated.
+
+=head2 Macros
+
+B<OPENSSL_VERSION_NUMBER> is a combination of the major, minor and
+patch version into a single integer 0xMNN00PP0L, where:
+
+=over 4
+
+=item M
+
+is the number from B<OPENSSL_VERSION_MAJOR>, in hexadecimal notation
+
+=item NN
+
+is the number from B<OPENSSL_VERSION_MINOR>, in hexadecimal notation
+
+=item PP
+
+is the number from B<OPENSSL_VERSION_PATCH>, in hexadecimal notation
+
+=back
+
+B<OPENSSL_VERSION_TEXT> is the text that combines the other macros
+roughly (it's a pseudo code) like this:
+
+ "OpenSSL " \
+ OPENSSL_VERSION_MAJOR "." OPENSSL_VERSION_MINOR "." OPENSSL_VERSION_PATCH \
+ OPENSSL_VERSION_PRE_RELEASE OPENSSL_VERSION_BUILD_METADATA \
+ " " \
+ OPENSSL_RELEASE_DATE
+
+=head2 Functions
+
+OpenSSL_version_num() returns the value of B<OPENSSL_VERSION_NUMBER>.
+
+OpenSSL_version() returns the same values as OPENSSL_info() for the same
```

Comment:
The motivation is that `OpenSSL_version()` probably was just about the version number and not about all the other information it provides, and since `OPENSSL_info()` doesn't serve the version number at all (because it's not one calculated number any more), the old name doesn't correspond at all to what it actually does.

Reply:
Well it provides information about this *version* of OpenSSL :-)

I can understand deprecating a function where the parameters or the return type is wrong. But deprecating and introducing a new version because the name isn't as descriptive as we would like kind of seems like breakage for breakage sake (ok we're not breaking anything yet since we're not removing the old function yet - but in time we will). It's not like the old name is actively misleading (as was the case in the rename of SSLv23_method to TLS_method).

Reply:
We disagree, I associate the name with what it returns, and I *do* find the name strongly misleading.

Reply:
Ok. I raised it on project and asked for more input.

Reply:
So how do we get the version number in the future? Pretty much everything that uses OpenSSL expects to be able to find a simple numeric string somewhere. HTTP User-Agent strings and various other tools want it.

Reply:
`OPENSSL_version_text()` and `OPENSSL_version_text_full()`, which are documented in this very manual.

Reply:
My two bits: The old name isn't great, but of the retrospectively not-great APIs in openssl that are a problem, this wouldn't even be on my list. I wouldn't delete the existing API.

Its not clear why this PR introduces a new `_info(t)` API that takes a number as an argument to select some bit of non-version information, but then introduces 7 new APIs to get various pieces of the openssl version. Why not just have an API per piece of information for version and non-version info? Of if less functions are an API goal, why not return a `const struct OpenSSL_info*`? That would be easy to get all the info out of, and could ABI-compatibly have things added to the end of the struct.

Reply:
5 of those new functions are to get the library's idea of the values for the corresponding 5 macros.  Remains two that could arguably be folded into returned strings by OPENSSL_info() or OpenSSL_version(), whichever we end up choosing, if that's more desirable.

I would rather not add a public structure...

Reply:
The current `OpenSSL_version(what)`, which returns strings describing various features of the run-time OpenSSL library is a reasonably intuitive interface, it is mostly used just with `what == OPENSSL_VERSION`, and the name fits the primary use-case.

In Postfix for example, you can query that value via "postconf -T run-version":
```
$ uname -sr
FreeBSD 11.2-RELEASE-p4
$ postconf -T run-version
OpenSSL 1.1.1a  20 Nov 2018
```
```
$ uname -sr
Darwin 18.2.0
$ postconf -T run-version
LibreSSL 2.2.7
```

We just recently (in the API changes from 1.0.2 to 1.1.0) renamed the function from `SSLeay_version`, and changing it again IMHO just punishes the users who've gone to the trouble of adapting to the preferred 1.1.0 interface.  The API reporting the version should not change with every API version!

We did too much cosmetic renaming in the transition from 1.0.2 to 1.1.0, let's not repeat that mistake. API stability trumps cosmetic issues.

Reply:
I've heard you all and had a think, and I think I may have come up with a solution that doesn't clash all over the place.  Fixup commits coming up